From b20b8f9b728797b27b0531b49755a0985a0a33fc Mon Sep 17 00:00:00 2001
From: Lu Feifei <lufeifei@wxiat.com>
Date: Tue, 22 Nov 2022 17:40:39 +0800
Subject: [PATCH] sw64: kvm: support debugging guest kernel via gdb
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: b20b8f9b728797b27b0531b49755a0985a0a33fc
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56WV8

--------------------------------

To support qemu-gdb, this patch adds a new exit type EXIT_DEBUG
to handle exits for breakpoint debugging.

Signed-off-by: Lu Feifei <lufeifei@wxiat.com>
Reviewed-by: He Sheng <hesheng@wxiat.com>
Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/include/asm/kvm_asm.h | 4 +++-
 arch/sw_64/kvm/handle_exit.c     | 4 ++++
 arch/sw_64/kvm/kvm-sw64.c        | 5 +++--
 3 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/arch/sw_64/include/asm/kvm_asm.h b/arch/sw_64/include/asm/kvm_asm.h
index 841bfa1dd0aa..30d3ccbabff0 100644
--- a/arch/sw_64/include/asm/kvm_asm.h
+++ b/arch/sw_64/include/asm/kvm_asm.h
@@ -12,6 +12,7 @@
 #define SW64_KVM_EXIT_IPI		14
 #define SW64_KVM_EXIT_RESTART		17
 #define SW64_KVM_EXIT_FATAL_ERROR	22
+#define SW64_KVM_EXIT_DEBUG		24
 
 #ifdef CONFIG_KVM_MEMHOTPLUG
 #define SW64_KVM_EXIT_MEMHOTPLUG	23
@@ -26,6 +27,7 @@
 	{14, "IPI" },		\
 	{17, "RESTART" },	\
 	{22, "FATAL_ERROR" },	\
-	{23, "MEMHOTPLUG" }
+	{23, "MEMHOTPLUG" },	\
+	{24, "DEBUG" }
 
 #endif /* _ASM_SW64_KVM_ASM_H */
diff --git a/arch/sw_64/kvm/handle_exit.c b/arch/sw_64/kvm/handle_exit.c
index 52f40a4c5803..8304ebfcd5c6 100644
--- a/arch/sw_64/kvm/handle_exit.c
+++ b/arch/sw_64/kvm/handle_exit.c
@@ -43,6 +43,10 @@ int handle_exit(struct kvm_vcpu *vcpu, struct kvm_run *run,
 	case SW64_KVM_EXIT_IPI:
 		vcpu_send_ipi(vcpu, hargs->arg0);
 		return 1;
+	case SW64_KVM_EXIT_DEBUG:
+		vcpu->run->exit_reason = KVM_EXIT_DEBUG;
+		vcpu->run->debug.arch.epc = vcpu->arch.regs.pc;
+		return 0;
 #ifdef CONFIG_KVM_MEMHOTPLUG
 	case SW64_KVM_EXIT_MEMHOTPLUG:
 		vcpu_mem_hotplug(vcpu, hargs->arg0);
diff --git a/arch/sw_64/kvm/kvm-sw64.c b/arch/sw_64/kvm/kvm-sw64.c
index 6d34bc843167..0f0fa9b586cc 100644
--- a/arch/sw_64/kvm/kvm-sw64.c
+++ b/arch/sw_64/kvm/kvm-sw64.c
@@ -562,9 +562,10 @@ int kvm_arch_vcpu_ioctl_get_regs(struct kvm_vcpu *vcpu, struct kvm_regs *regs)
 	return 0;
 }
 
-int kvm_arch_vcpu_ioctl_set_guest_debug(struct kvm_vcpu *vcpu, struct kvm_guest_debug *dbg)
+int kvm_arch_vcpu_ioctl_set_guest_debug(struct kvm_vcpu *vcpu,
+						struct kvm_guest_debug *dbg)
 {
-	return -ENOIOCTLCMD;
+	return 0;
 }
 
 void _debug_printk_vcpu(struct kvm_vcpu *vcpu)
-- 
2.33.0

